src/main/scala/mesosphere/marathon/core/health/impl/HealthCheckWorker.scala

```diff
@@ -3,126 +3,122 @@ package core.health.impl
 import akka.http.scaladsl.model.{ HttpRequest, HttpResponse, headers }
 import akka.http.scaladsl.settings.ClientConnectionSettings
 import akka.http.scaladsl.{ ConnectionContext, Http }
+import akka.stream.ActorMaterializer
 import akka.stream.scaladsl.{ Sink, Source }
-import akka.stream.{ ActorMaterializer, ActorMaterializerSettings, Materializer }
 import com.typesafe.scalalogging.StrictLogging
 import com.typesafe.sslconfig.akka.AkkaSSLConfig
+import javax.net.ssl.{ KeyManager, SSLContext, X509TrustManager }
+import mesosphere.marathon.Protos.HealthCheckDefinition.Protocol
 import mesosphere.marathon.core.health._
 import mesosphere.marathon.core.instance.Instance
 import mesosphere.marathon.state.{ AppDefinition, Timestamp }
 import mesosphere.util.ThreadPoolContext
 
-import scala.concurrent.Future
 import scala.concurrent.duration.FiniteDuration
+import scala.concurrent.{ ExecutionContext, Future, Promise }
 import scala.util.control.NonFatal
-import scala.util.{ Failure, Success }
-
-class HealthCheckWorkerActor(implicit mat: Materializer) extends Actor with StrictLogging {
-
-  import HealthCheckWorker._
-
-  private implicit val system = context.system
-  private implicit val scheduler = system.scheduler
-  import context.dispatcher // execution context for futures
-  private implicit val materializer = ActorMaterializer(ActorMaterializerSettings(system))
-
-  def receive: Receive = {
-    case HealthCheckJob(app, instance, check) =>
-      val replyTo = sender() // avoids closing over the volatile sender ref
-
-      doCheck(app, instance, check)
-        .andThen {
-          case Success(Some(result)) => replyTo ! result
-          case Success(None) => // ignore
-          case Failure(t) =>
-            logger.warn(s"Performing health check for app=${app.id} instance=${instance.instanceId} port=${check.port} failed with exception", t)
-            replyTo ! Unhealthy(
-              instance.instanceId,
-              instance.runSpecVersion,
-              s"${t.getClass.getSimpleName}: ${t.getMessage}"
-            )
-        }
-        .onComplete { _ => self ! PoisonPill }
+import scala.util.{ Failure, Success, Try }
+
+object HealthCheckWorker extends StrictLogging {
+
+  implicit class RichFuture[T](f: Future[T])(implicit ec: ExecutionContext) {
+    def mapAll[U](pf: PartialFunction[Try[T], U]): Future[U] = {
+      val p = Promise[U]()
+      f.onComplete(r => p.complete(Try(pf(r))))
+      p.future
+    }
   }
 
-  def doCheck(
-    app: AppDefinition, instance: Instance, check: MarathonHealthCheck): Future[Option[HealthResult]] = {
+  def run(app: AppDefinition, instance: Instance, healthCheck: MarathonHealthCheck)(implicit mat: ActorMaterializer): Future[HealthResult] = {
+    logger.debug("Dispatching health check job for {}", instance.instanceId)
+
+    implicit val system = mat.system
+    implicit val ex = mat.executionContext
+
+    check(app, instance, healthCheck)
+      .mapAll {
```

Comment:
Shouldn't this work for a `map` as well?

Reply:
How that? `map` is for successful execution, `mapAll` is only needed because Scala 2.11 doesn't have `transform` (introduced in 2.12)

Reply:
Ah, got it.